File: backend/src/main/java/iuh/labbooking/service/slot/SlotService.java
package iuh.labbooking.service.slot;

import iuh.labbooking.dto.request.slot.BulkSlotStatusRequest;
import iuh.labbooking.dto.request.slot.SlotRequest;
import iuh.labbooking.dto.response.base.PageResponse;
import iuh.labbooking.dto.response.slot.SlotResponse;
import iuh.labbooking.dto.response.slot.SecureSlotResponse;

import java.util.List;

public interface SlotService {
    SecureSlotResponse createSlot(SlotRequest request);

    SecureSlotResponse updateSlot(Long id, SlotRequest request);

    SecureSlotResponse findSlotById(Long id);

    PageResponse<List<SlotResponse>> findSlots(int page, int limit, String keyword);

    PageResponse<List<SecureSlotResponse>> filterSlots(int page, int limit, String keyword, Boolean active);

    void toggleSlotStatus(Long id);
    
    void updateSlotsStatus(BulkSlotStatusRequest request);
}

File: frontend/src/services/slot.service.ts
import http from '@/lib/http'
import { type ApiResponse, type PageResponse } from '@/schemas/base.schema'
import { type SecureSlotResponse, type SlotRequest, type SlotResponse } from '@/schemas/slot.schema'

export const slotService = {
  getSlots: (params: { page: number; limit: number; keyword?: string }) =>
    http.get<ApiResponse<PageResponse<SlotResponse[]>>>('/slots', { params }),

  getSlotById: (id: number) => http.get<ApiResponse<SecureSlotResponse>>(`/slots/${id}`),

  createSlot: (data: SlotRequest) => http.post<ApiResponse<SecureSlotResponse>>('/slots', data),

  updateSlot: (id: number, data: SlotRequest) => http.put<ApiResponse<SecureSlotResponse>>(`/slots/${id}`, data),

  filterSlots: (params: { page: number; limit: number; keyword?: string; active?: boolean }) =>
    http.get<ApiResponse<PageResponse<SecureSlotResponse[]>>>('/slots/admin', { params }),

  toggleActive: (id: number) => http.patch<ApiResponse<void>>(`/slots/${id}/status`),

  updateBulkStatus: (ids: number[], active: boolean) =>
    http.patch<ApiResponse<void>>('/slots/actions/bulk-active', { ids, active })
}
